### PR TITLE
ci: add fail-closed delta routing and cut full PR CI to six minutes

### PR DESCRIPTION
**Pull request by Bojan131**
## Why

Pull requests currently fan out across almost the entire monorepo test matrix, even when they change only documentation or one isolated package. Ten recent successful PR runs had a 14-minute median, and the same full workflow on this PR originally took 13m07s.

## What changed

- Adds a deterministic, fail-closed planner that maps PR changes to owning and downstream Node lanes plus affected EVM integration scopes.
- Keeps full CI for merge-queue candidates, protected-branch pushes, manual runs, global or unknown files, manifests, high-risk packages, large/ambiguous diffs, a `ci:full` override, and a deterministic 5% audit sample.
- Adds stable `CI gate` and `EVM integration gate` jobs that fail if any selected lane is skipped, fails, or is cancelled.
- Compiles the two EVM test configurations once and shares immutable artifacts with their consumers.
- Splits the long full-CI bottlenecks across isolated runners: CLI into four duration-balanced shards, chain into three duration-balanced shards, and all 326 real-node Playwright tests into seven separate four-node devnets.
- Discovers CLI and chain test files at runtime and tests the shard planner for exact union, no overlap, deterministic output, and conservative handling of new files.
- Runs sharded Solidity tests on PR and merge-queue candidates while retaining the full coverage ratchet on protected pushes and manual runs.
- Adds previously omitted RDF utilities, OKF, and demo tests; the demo command includes both Kafka Streams and EPCIS Bike suites.
- Stabilizes two pre-existing false failures exposed by full validation: the Node UI test now waits for its lazy Shiki import, and the chain retry-budget test uses deterministic fake timers.

This avoids introducing a `develop` branch: PRs get fast feedback, while GitHub's merge queue validates the exact combined candidate before merge. Snapshot-style tests guard the routing decisions and dependency closure.

## Safety and rollout

Delta routing is deliberately **disabled by default**. Until repository administrators set `CI_DELTA_ENABLED=true`, PRs continue to exercise full CI; the live benchmark below therefore validates the complete path, not a selective shortcut.

Before enabling the variable:

1. Create the `ci:full` label.
2. Enable merge queue for `main`.
3. Require `CI gate` and `EVM integration gate`.
4. Set `CI_DELTA_ENABLED=true`.

Every matrix shard uses its own runner and isolated Hardhat/devnet state. Unknown paths fail closed to full CI, and the aggregate gate checks both shared artifact producers and every selected matrix job.

Forced-full parallel CI trades compute for latency: the measured runner total increased from 87.9 to 109.0 minutes. Delta routing reduces runner use on ordinary leaf/docs PRs; global and high-risk changes pay the extra parallel compute for near-six-minute feedback. The unsharded Solidity coverage ratchet remains a roughly 25-minute push/manual safety net and is outside the PR feedback path.

## Validation

- Full live GitHub run: [6m05s, 46 successful jobs, one policy-skipped coverage job, zero failures](https://github.com/OriginTrail/dkg/actions/runs/29325452377), down from [13m07s](https://github.com/OriginTrail/dkg/actions/runs/29321182804) on the same PR (53.5% faster).
- CLI: 2,705 passed + 12 skipped = 2,717 tests across all 190 files; four shards resolved exactly once.
- Chain: 1,033 passed + 1 skipped = 1,034 tests across all 58 active files; three shards resolved exactly once.
- Real-node Playwright: 318 passed + 8 skipped = all 326 tests across seven isolated four-node devnets.
- `pnpm test:scripts`: 86/86 passed, including planner, exact-coverage, and aggregate-gate tests.
- `actionlint`, Zizmor auditor, and `git diff --check`: passed; Zizmor reported no findings.
- `DKG_SKIP_EVM_BUILD=1 pnpm run build:packages`: 22/22 tasks passed in the earlier full validation.

The workflow documentation records the measured timings, retained coverage, rollout controls, and the latency/compute trade-off.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -38,56 +40,93 @@ permissions:
...
+            # unrelated changes that landed on base as part of this PR.
+            git diff --name-status -z "${BASE_SHA}" "${MERGE_SHA}" > "${CHANGES_FILE}"
+          fi
+
+          PLAN_EVENT="${EVENT_NAME}"
+          if [[ "${EVENT_NAME}" == "pull_request" && "${DELTA_ENABLED}" != "true" ]]; then
+            PLAN_EVENT="pull_request_delta_disabled"
+          fi
+
+          node scripts/ci/plan-ci.mjs \
```
Comment (otReviewAgent) [inline .github/workflows/ci.yml]: 🔴 Bug: Delta CI trusts PR-controlled planner and gate code

**What's wrong**
The new fail-closed rules are implemented inside repository scripts that are executed from the checked-out PR merge commit. That means changes to the planner, result validator, or their shared libraries can change the rules that decide whether those same changes run full CI, and can also change the aggregate gate that branch protection is meant to trust. The `scripts/**` full-CI rule does not protect against this because it is evaluated by the PR-controlled planner itself.

**Example**
A PR changes `scripts/ci/plan-ci.mjs` to emit `run_node=false`, every lane as `false`, and `evm_matrix=[]`, then changes `scripts/ci/assert-ci-results.mjs` to exit 0. The workflow uses those PR-controlled files, so all lanes are skipped and the documented aggregate gate can still pass.

**Suggested direction**
Do not let the merge candidate's copy of the planner or gate be the authority for whether that same candidate runs tests. Use a trusted/base-ref copy, a pinned external action, or workflow-shell logic that forces full CI for workflow/planner/gate changes before invoking repository code.

**For Agents**
Look at the planner and gate invocation in both workflows. Preserve delta CI for ordinary package changes, but make planner/gate decisions come from a trusted source or add a workflow-level fail-closed precheck for `.github/workflows/**` and `scripts/**` before any PR-controlled planner code runs. Add a test or CI fixture showing that a PR touching the planner/gate/workflow paths cannot produce a skipped full-CI plan.

File: scripts/lib/ci-delta.mjs [hunk 2]
```diff
@@ -0,0 +1,506 @@
...
+  'kosava_node_ui',
+  'kosava_node_ui_e2e',
+  'kosava_supporting',
+  'kosava_hardhat_plugins',
+  'contracts',
+]);
+
+export const EVM_SCOPES = Object.freeze(['chain', 'publisher', 'agent']);
+
+export const WORKSPACE_RULES = Object.freeze({
```
Comment (otReviewAgent) [inline scripts/lib/ci-delta.mjs:18]: 🟡 Issue: Replace the hand-expanded routing matrix with a derived model

**What's wrong**
The new planner encodes the same dependency-routing knowledge as large copied arrays. That preserves too much incidental state and makes the CI policy harder to audit: the tests verify that the duplicated table is complete, but the implementation still relies on humans keeping the expanded table in sync with the package graph.

**Example**
`packages/core` and `packages/rdf-utils` repeat the same ten-lane array, while other shared packages repeat overlapping suffixes. Adding a new consumer lane means knowing to update every upstream provider's expanded list, plus the owner/job maps, rather than updating one canonical dependency model.

**Suggested direction**
Make direct lane ownership the canonical map, derive downstream consumers algorithmically from package manifests, and keep exceptional integration edges in a small declarative policy table. That would delete most of `WORKSPACE_RULES` and make future package/lane changes local.

**For Agents**
In `scripts/lib/ci-delta.mjs`, preserve the current selected-lane snapshots but replace the hand-expanded downstream lists with direct owner metadata plus derived reverse-dependency closure from workspace manifests. Keep only explicit integration overrides and `forceFull` as policy data.

File: .github/workflows/evm-integration.yml [hunk 3]
```diff
@@ -3,50 +3,87 @@ name: EVM Integration Tests
...
+    timeout-minutes: 5
+    outputs:
+      evm_matrix: ${{ steps.plan.outputs.evm_matrix }}
+      plan_json: ${{ steps.plan.outputs.plan_json }}
+    steps:
+      - uses: actions/checkout@9c091bb21b7c1c1d1991bb908d89e4e9dddfe3e0 # v7.0.0
+        with:
+          persist-credentials: false
+          fetch-depth: 2
+      - name: Compute fail-closed EVM plan
```
Comment (otReviewAgent) [inline .github/workflows/evm-integration.yml:50]: 🟡 Issue: Centralize the duplicated delta-planning workflow wrapper

**What's wrong**
The planning algorithm is extracted to JS, but the fragile orchestration around it is copied in two YAML files. That keeps the most environment-sensitive part of the implementation split across workflows, which is exactly where drift is hardest to review.

**Example**
Both workflow blocks create `CHANGES_FILE`, fetch `BASE_SHA`, diff `${BASE_SHA}` to `${MERGE_SHA}`, rewrite the event to `pull_request_delta_disabled`, then call `node scripts/ci/plan-ci.mjs`. A future fix to the diff boundary or rollout gate has to be applied twice or the two workflows can diverge.

**Suggested direction**
Move the GitHub-specific wrapper around `plan-ci.mjs` into a shared entrypoint, for example `scripts/ci/plan-from-github-env.mjs` or a composite action. Both workflows should call the same wrapper rather than carrying copied shell policy.

**For Agents**
Extract the duplicated GitHub-context orchestration into one script or local composite action. Preserve the current outputs and summaries, and leave each workflow responsible only for naming its job and exposing the outputs it needs.

File: scripts/lib/__tests__/ci-delta.test.mjs [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+      workflow.includes(`needs.changes.outputs.${lane} == 'true'`),
+      `${job} must be gated by ${lane}`,
+    );
+  }
+  assert.ok(workflow.includes("needs.changes.outputs.contracts == 'true'"));
+  assert.ok(
+    workflow.includes('run: node scripts/check-npm-metadata.mjs'),
+    'docs-only package README changes must retain the npm metadata gate',
+  );
+  assert.ok(
```
Comment (otReviewAgent) [inline scripts/lib/__tests__/ci-delta.test.mjs:434]: 🟡 Issue: Default-off rollout behavior is only string-checked, not behaviorally verified

**What's wrong**
The PR relies on CI_DELTA_ENABLED being default-off as the main safety control for selective CI, but the test only greps for the variable and base-ref expression. That gives weak confidence because it does not prove the disabled path actually produces a full plan.

**Example**
If the PLAN_EVENT rewrite were removed but the DELTA_ENABLED env expression stayed in the workflow, this test would still pass, while PRs with CI_DELTA_ENABLED unset could run delta CI instead of full CI.

**Suggested direction**
Cover the synthetic disabled event and its workflow wiring with a test that would fail if the repository variable stopped forcing full CI.

**For Agents**
Add behavioral coverage for the default-off path. At minimum, test that planCi or plan-ci.mjs with event pull_request_delta_disabled returns full CI, and tighten the workflow wiring test so both workflows must pass that synthetic event when DELTA_ENABLED is not true.

File: scripts/lib/ci-delta.mjs (hunk 2 above)
Comment (otReviewAgent) [reply scripts/lib/ci-delta.mjs:18]: 🟡 Issue: Derive routing from ownership/dependencies instead of hand-maintaining downstream closures

**What's wrong**
This PR introduces a large manual routing manifest that encodes computed state rather than source-of-truth state. The same dependency closure is then recomputed in tests, which catches some drift but still leaves the production planner as a sprawling N-by-M table that must be manually edited whenever package relationships change.

**Example**
`packages/core` and `packages/rdf-utils` duplicate the same long lane list, and several packages repeat the same `bura_cli`/`kosava_node_ui_e2e`/`kosava_hardhat_plugins` downstream set. Adding one new workspace consumer means updating every affected upstream rule instead of one direct ownership/dependency relationship.

**Suggested direction**
Make the planner compute the transitive downstream closure from workspace manifests and `WORKSPACE_OWNING_LANES`, with a small explicit overlay for non-package integration lanes. That deletes most of the table and turns future dependency changes into one local update.

**For Agents**
In `scripts/lib/ci-delta.mjs`, preserve the current selected-lane behavior but replace the hand-expanded downstream closures with a model that stores direct test owners plus explicit integration overlays/force-full rules. Reuse the package manifest reverse-dependency walk currently embedded in `scripts/lib/__tests__/ci-delta.test.mjs`, then keep snapshot tests proving the resulting plans are unchanged.

File: .github/workflows/evm-integration.yml (hunk 3 above)
Comment (otReviewAgent) [reply .github/workflows/evm-integration.yml:50]: 🟡 Issue: The planner bootstrap shell is duplicated across workflows

**What's wrong**
The PR adds a second copy of the most delicate part of selective CI: collecting the changed-file input and translating rollout state into the event passed to the planner. That is avoidable orchestration duplication in YAML, where drift is hard to review and refactor safely.

**Example**
If the diff base changes from `BASE_SHA`/`MERGE_SHA` to another comparison, or the `CI_DELTA_ENABLED` rollout condition changes, both workflow YAML blocks must be edited identically. Any drift makes primary CI and EVM integration plan from different inputs.

**Suggested direction**
Collapse this into one reusable planner invocation path so workflow YAML only declares env/output wiring. The code that decides what diff to read and whether delta is enabled should have a single implementation.

**For Agents**
Move the shared pull-request diff acquisition and delta-enabled event normalization into `scripts/ci/plan-ci.mjs` or a small checked-in shell/composite helper. Keep both workflows passing the same env values and assert one shared entry point in the script tests.

File: scripts/lib/ci-results.mjs [hunk 5]
```diff
@@ -0,0 +1,112 @@
+import { CI_LANES, EVM_SCOPES } from './ci-delta.mjs';
+
+export const PRIMARY_LANE_JOBS = Object.freeze({
```
Comment (otReviewAgent) [inline scripts/lib/ci-results.mjs:3]: 🟡 Issue: Lane metadata is split across parallel registries and special cases

**What's wrong**
The selective-CI model now requires several independent structures to agree about what a lane is, which job owns it, whether it needs the shared Node build, and which lanes are special. That makes adding or changing a lane more error-prone than necessary and encourages more ad-hoc branches as the matrix grows.

**Example**
`bura_blazegraph_arm64` has to be excluded from `NODE_LANES` in the planner and excluded again from `selectedNodeLane` in the result gate. `contracts` is in `CI_LANES` but not in `PRIMARY_LANE_JOBS`, so it needs separate validation branches.

**Suggested direction**
Centralize lane metadata in one typed-ish registry and derive the planner outputs and gate checks from it. That would remove the repeated filters and the separate special-case handling for arm64 and contracts lanes.

**Confidence note**
This is a structural maintainability concern rather than a current behavior defect; the exact refactor shape may depend on how much workflow YAML the project is willing to generate or validate from JS.

**For Agents**
Introduce a single lane definition table with fields such as `id`, `primaryJob`, `requiresNodeBuild`, and `kind`. Derive `CI_LANES`, `NODE_LANES`, `PRIMARY_LANE_JOBS`, GitHub outputs, and gate validation from that table, preserving current workflow job names and current gate behavior.

File: scripts/lib/__tests__/ci-delta.test.mjs [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+    'bura_query',
+    'kosava_node_ui',
+    'kosava_node_ui_e2e',
+    'kosava_supporting',
+    'kosava_hardhat_plugins',
+  ]);
+  assert.deepEqual(core.evmScopes, EVM_SCOPES);
+});
+
+test('highest-risk, global, unknown, manifest, large, and ambiguous changes fail closed', () => {
```
Comment (otReviewAgent) [inline scripts/lib/__tests__/ci-delta.test.mjs:132]: 🟡 Issue: Fail-closed edge cases are only partially covered

**What's wrong**
The new delta planner relies on failing closed when Git reports ambiguous changes or no changed files, but the unit test only proves one ambiguous status (`D`). Rename/copy and no-diff paths are explicitly part of the rollout safety policy, so leaving them untested weakens confidence in the selective-CI gate.

**Example**
A test like `pullRequestPlan([{ status: 'R100', paths: ['packages/agent/src/a.ts', 'packages/agent/src/b.ts'] }])` should assert `fullCi === true`, and `pullRequestPlan([])` should do the same. Today those regressions could pass if only the untested branches changed.

**Suggested direction**
Parameterize the fail-closed planner tests over the risky git status letters and add a no-diff case so the selective-CI safety contract is regression-tested.

**For Agents**
Update `scripts/lib/__tests__/ci-delta.test.mjs`; extend the fail-closed table around the highest-risk/ambiguous test to cover `R100`, `C100`, and the empty change list, preserving the existing full-CI behavior for deletion, manifests, unknown paths, large PRs, and workspace fanout.

File: scripts/lib/ci-delta.mjs (hunk 2 above)
Comment (otReviewAgent) [reply scripts/lib/ci-delta.mjs:18]: 🟡 Issue: Derive the routing table instead of hand-maintaining the downstream closure

**What's wrong**
The new planner stores expanded downstream lane closures by hand, then also stores direct owner mappings and tests that recompute the closure. That creates multiple sources of truth for the same routing model. The structure is harder to audit than necessary and invites drift as packages and dependency edges change.

**Example**
packages/core and packages/rdf-utils both spell out the same 10-lane downstream list, while packages/storage, packages/query, packages/publisher, and packages/agent each repeat overlapping subsets. Adding one dependency edge or lane owner now requires keeping WORKSPACE_RULES, WORKSPACE_OWNING_LANES, WORKSPACE_OWNING_EVM_SCOPES, and tests aligned.

**Suggested direction**
Use the reverse-dependency graph as the canonical model in the planner, then layer explicit policy exceptions on top. That deletes most repeated lane arrays and makes future routing changes local to the owner/override declaration instead of a manually expanded closure.

**For Agents**
Look in scripts/lib/ci-delta.mjs. Preserve the current selected lanes/scopes, but derive the downstream closure from package manifests plus WORKSPACE_OWNING_LANES/WORKSPACE_OWNING_EVM_SCOPES. Keep only small explicit overlays for undeclared integrations, blazegraph arm64, and force-full workspaces. Update scripts/lib/__tests__/ci-delta.test.mjs to assert the derived snapshots stay unchanged.

File: .github/workflows/evm-integration.yml (hunk 3 above)
Comment (otReviewAgent) [reply .github/workflows/evm-integration.yml:50]: 🟡 Issue: Centralize the repeated planner orchestration shared by both workflows

**What's wrong**
The PR introduces a good planner abstraction, but leaves a substantial part of the planning algorithm duplicated in YAML. That duplication is not just boilerplate; it encodes the fail-closed diff source, rollout gate, and event remapping policy. Those are core invariants and should have one implementation.

**Example**
A future change from comparing BASE_SHA..MERGE_SHA to another diff source, or changing the delta rollout predicate, must be made in both workflows. If one workflow misses the update, the primary and EVM gates can plan different CI behavior for the same PR.

**Suggested direction**
Let the JavaScript planner own diff collection and delta rollout normalization, or wrap this block in a single reusable workflow/composite action. The YAML should pass GitHub context through once rather than duplicating the control flow in every workflow that needs a plan.

**For Agents**
Move the repeated shell into scripts/ci/plan-ci.mjs or a small shared CI helper/composite action. Preserve the current inputs: EVENT_NAME, BASE_SHA, MERGE_SHA, labels, sample key, delta-enabled rollout, GitHub output, and summary. Both workflows should call the same helper and only declare the outputs they consume.

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -17,6 +14,11 @@ on:
     branches: [main, v10-rc, release/rc.12, rc17-vm-wip]
   pull_request:
     branches: [main, v10-rc, release/rc.12, rc17-vm-wip]
+    types: [opened, synchronize, reopened, labeled, unlabeled]
```
Comment (otReviewAgent) [inline .github/workflows/ci.yml:17]: 🟡 Issue: Workflow rewrites are not backed by workflow syntax validation

**What's wrong**
The changed behavior depends on GitHub Actions YAML and expression semantics, but the added tests only exercise the JavaScript planner/gate helpers and string-match selected workflow snippets. Without actionlint or an equivalent workflow validation run, this PR has no direct evidence that the rewritten workflows are syntactically valid and loadable by GitHub Actions.

**Example**
A malformed or unsupported workflow expression in one of the new job conditions could still satisfy the current string checks, while GitHub Actions rejects or mis-evaluates the workflow before the new CI gate can protect anything.

**Suggested direction**
Add actionlint or equivalent workflow validation to the verification path, then report that result for this PR.

**Confidence note**
This is based on the supplied verification notes. If actionlint or an equivalent GitHub workflow validation already ran in CI for this PR, this can be dismissed.

**For Agents**
Run actionlint against .github/workflows/ci.yml and .github/workflows/evm-integration.yml, and preferably add a small CI/local test target that lints those workflows whenever the delta planner or workflow wiring changes. Keep the existing planner behavior tests as-is; this should validate the YAML/expression layer they do not parse.

File: .github/workflows/evm-integration.yml (hunk 3 above)
Comment (otReviewAgent) [reply .github/workflows/evm-integration.yml:50]: 🟡 Issue: Collapse the duplicated planner bootstrap into one tested path

**What's wrong**
This PR makes CI planning a control-plane feature, but the workflow input collection is copy-pasted across the primary and EVM workflows. That spreads policy-sensitive logic across YAML blocks, making future changes brittle and harder to review.

**Example**
Changing the main-only rollout expression, merge diff strategy, label handling, or delta-disabled mode now requires keeping two YAML shell blocks in sync.

**Suggested direction**
Move the GitHub-event bootstrap into the JS planner or one shared wrapper so each workflow only supplies workflow-specific context and consumes outputs. Keep CI policy logic in the tested script layer instead of duplicating shell orchestration in YAML.

**For Agents**
Refactor both workflow plan jobs so the shared PR-diff/bootstrap logic lives in scripts/ci/plan-ci.mjs or a small shared CI script/composite action. Preserve BASE_SHA/MERGE_SHA diff semantics, DELTA_ENABLED behavior, labels/sample-key handling, GitHub outputs, and summary output; add a focused script test for enabled and delta-disabled PR inputs.

File: scripts/lib/ci-delta.mjs (hunk 2 above)
Comment (otReviewAgent) [reply scripts/lib/ci-delta.mjs:18]: 🟡 Issue: Derive CI routing instead of hand-expanding every downstream lane

**What's wrong**
The planner encodes the dependency graph as large hand-maintained lane lists, then adds separate ownership maps and tests that partially reconstruct the same graph. That is a lot of incidental structure for a routing policy and makes the next package/dependency change harder than it needs to be.

**Example**
packages/core and packages/rdf-utils carry identical long lane lists, while storage/chain/query/publisher repeat near-variants. Adding a workspace or dependency edge now means updating multiple maps plus tests, even though the package manifests already describe most of the graph.

**Suggested direction**
Use one canonical lane/workspace model and compute the transitive downstream lanes at plan time. Keep only the irreducible policy data as declarations; delete the repeated expanded arrays where possible.

**For Agents**
In scripts/lib/ci-delta.mjs, replace the fully expanded per-workspace lane arrays with a smaller manifest: owner lanes, explicit integration/undeclared edges, forceFull flags, and EVM owners. Derive downstream lane closure from package.json reverse dependencies, then compare current snapshot outputs to preserve behavior.

File: scripts/ci/plan-vitest-shard.mjs [hunk 8]
```diff
@@ -0,0 +1,451 @@
...
+ */
+import { readdirSync } from 'node:fs';
+import path from 'node:path';
+import process from 'node:process';
+import { fileURLToPath } from 'node:url';
+
+export const PER_FILE_OVERHEAD_MS = 600;
+export const UNKNOWN_FILE_BODY_MS = 60_000;
+
+const CLI_BODY_MS = Object.freeze({
```
Comment (otReviewAgent) [inline scripts/ci/plan-vitest-shard.mjs:30]: 🟡 Issue: Separate shard timing data from the sharding engine

**What's wrong**
The new script mixes volatile historical timing data with reusable bin-packing logic, and it only handles cli/chain while the agent lane keeps a parallel greedy sharder. This creates unnecessary indirection and duplicate maintenance rules for the same CI concept.

**Example**
Refreshing timings for cli/chain means editing scripts/ci/plan-vitest-shard.mjs, while refreshing agent timings still means editing packages/agent/scripts/ci-shard-agent.mjs. The bin-packing idea is shared, but the implementation and data are split.

**Suggested direction**
Make the planner data-driven: a generic allocator plus per-package config/weights. That would remove the half-generic script shape and avoid another package-specific sharding mechanism living beside the existing agent sharder.

**For Agents**
Extract the shared shard allocator to scripts/lib and move per-package shard config/timing data into package-local or central JSON fixtures. Either migrate the agent sharder to the same engine or keep all package-specific sharding package-local. Preserve current file assignments, unknown-file weighting, excluded prefixes, and CLI stdout behavior with snapshot-style tests.

File: scripts/lib/ci-delta.mjs (hunk 2 above)
Comment (otReviewAgent) [reply scripts/lib/ci-delta.mjs:18]: 🟡 Issue: Derive lane routing instead of hard-coding the full closure per workspace

**What's wrong**
The planner’s core routing table is a manually expanded closure of every workspace’s downstream lanes. That makes the change larger than it needs to be and creates a maintenance trap: the real model is dependency ownership plus policy edges, but future maintainers have to edit many repeated arrays and rely on tests to catch drift after the fact.

**Example**
Adding one new downstream consumer lane for `packages/query` would require touching several expanded arrays such as `packages/core`, `packages/rdf-utils`, `packages/storage`, `packages/chain`, `packages/query`, and `packages/publisher`, instead of adding one direct owner or policy edge.

**Suggested direction**
Make `WORKSPACE_RULES` describe only direct ownership and explicit policy exceptions, then compute the transitive downstream lane closure in one place. That would delete most of the repeated arrays and make future lane/package changes local.

**For Agents**
Refactor `scripts/lib/ci-delta.mjs` so `planCi` preserves the current selected lanes/scopes but derives them from a smaller model: workspace owner lanes, package-manifest reverse dependencies, explicit integration edges, and force-full flags. Keep the existing routing snapshots as behavioral proof.

File: .github/workflows/evm-integration.yml (hunk 3 above)
Comment (otReviewAgent) [reply .github/workflows/evm-integration.yml:50]: 🟡 Issue: Pull the duplicated planner bootstrap out of YAML

**What's wrong**
Both workflows now carry a near-identical multi-step shell implementation for computing the changed-file input and invoking the planner. This is orchestration logic with real policy embedded in it, not workflow-specific YAML, so duplicating it makes the CI design harder to change safely.

**Example**
If the merge-commit diff strategy or rollout gate changes, both `.github/workflows/ci.yml` and `.github/workflows/evm-integration.yml` need the same shell edit. Missing one copy leaves two planners with subtly different inputs.

**Suggested direction**
Keep the workflows declarative and put the common PR diff preparation plus planner invocation behind one script entrypoint. The workflows should differ only in which outputs they expose and which gate validates them.

**For Agents**
Move the shared diff/bootstrap logic into a script such as `scripts/ci/write-ci-plan.mjs` or a small shell wrapper. Preserve current workflow outputs and summaries; update both workflows to pass env/context and call the single helper.

File: scripts/ci/plan-vitest-shard.mjs (hunk 8 above)
Comment (otReviewAgent) [reply scripts/ci/plan-vitest-shard.mjs:30]: 🟡 Issue: Separate shard timing data from the planner code

**What's wrong**
The shard planner mixes executable logic with hundreds of lines of static per-file duration data. The result is a 451-line script where the actual algorithm is buried under snapshot contents, making future changes harder to review and increasing the chance that timing maintenance touches code unnecessarily.

**Example**
Rebalancing shards after one timing refresh means editing a long source file full of raw snapshot data, even though the algorithm starts much later and is otherwise small.

**Suggested direction**
Store the timing snapshots as data files and have the script load and validate them. That keeps the algorithm readable and makes timing refreshes mechanical rather than source-code churn.

**For Agents**
Move timing data to versioned JSON fixtures under `scripts/ci/` with metadata for source run/job and package. Keep `plan-vitest-shard.mjs` focused on loading, validating, discovering files, and planning shards; preserve current stdout behavior and shard snapshots.

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -38,56 +40,93 @@ permissions:
...
+            fi
+            git cat-file -e "${MERGE_SHA}^{commit}"
+            # Compare the latest base with GitHub's synthetic merge candidate.
+            # Comparing base directly to a behind PR head would misclassify
+            # unrelated changes that landed on base as part of this PR.
+            git diff --name-status -z "${BASE_SHA}" "${MERGE_SHA}" > "${CHANGES_FILE}"
+          fi
+
+          PLAN_EVENT="${EVENT_NAME}"
+          if [[ "${EVENT_NAME}" == "pull_request" && "${DELTA_ENABLED}" != "true" ]]; then
```
Comment (otReviewAgent) [inline .github/workflows/ci.yml:127]: 🟡 Issue: Do not overload `eventName` with rollout state

**What's wrong**
The workflow encodes “delta is disabled” by inventing a fake event name. That is a stringly control channel across the workflow/script boundary, and it obscures the planner contract because event identity and rollout policy are separate concepts.

**Example**
A caller reading `planCi({ eventName })` has to know that `pull_request_delta_disabled` is a private control mode, while the aggregate gate later validates the same plan with the real `EVENT_NAME=pull_request`.

**Suggested direction**
Pass delta rollout as its own typed/explicit planner option and leave `eventName` as the real event. This makes the event boundary cleaner and avoids private pseudo-events leaking across the workflow/script API.

**Confidence note**
The sentinel currently works because `planCi` treats every non-`pull_request` event as full CI, but the boundary is still harder to reason about than necessary.

**For Agents**
Add an explicit planner input such as `deltaEnabled` or `forceFullReason` to `scripts/ci/plan-ci.mjs` and `planCi`. Keep `eventName` limited to real GitHub events and preserve current full-CI behavior when delta is disabled.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (Bojan131) [reply .github/workflows/ci.yml]: Fixed in f310533e8b29974d6809320f9c1af13970f98735.

Both workflows now treat the merge candidate only as diff input under `candidate/`; their planner and aggregate gate execute from `OriginTrail/dkg` pinned to immutable controller commit `c441bd6766ace67e331c0dfc6e22cb1325a6e1f6`. The candidate npm-metadata script also runs only after trusted plan outputs are finalized.

Regression coverage now recreates planner/gate edits that try to make every lane skip, proves workflow/planner/gate/shared-library changes select full CI, and proves both trusted gates reject all-skipped results. Local evidence: 90/90 script tests, actionlint, and zizmor passed. Live evidence: [full primary CI + trusted gate](https://github.com/OriginTrail/dkg/actions/runs/29327831858) passed in 6m, and [all EVM scopes + trusted gate](https://github.com/OriginTrail/dkg/actions/runs/29327831854) passed in 3m15s.

This closes the reported PR-controlled planner/gate-script bypass; normal `pull_request` workflow YAML itself remains protected by review/CODEOWNERS.

File: scripts/lib/ci-delta.mjs (hunk 2 above)
Comment (otReviewAgent) [reply scripts/lib/ci-delta.mjs:18]: 🟡 Issue: Collapse the CI routing table into one source of truth

**What's wrong**
The planner introduces a second dependency graph in code. WORKSPACE_RULES contains repeated, expanded downstream lane lists, while WORKSPACE_OWNING_LANES repeats the direct ownership model used to validate those lists. That makes the routing policy harder to audit and update because the reader has to reconcile multiple tables that describe the same architecture.

**Example**
A new package dependency edge now requires touching at least the workspace rule, the owning-lane table, and the routing tests. The tests can detect some drift, but the production planner still depends on a manually duplicated closure rather than a single model of package ownership and downstream consumers.

**Suggested direction**
Use a single declarative model plus computed closures instead of maintaining both full downstream lane lists and direct owner maps by hand. Keep explicit overrides for undeclared integration relationships, but make the common case derived.

**For Agents**
In scripts/lib/ci-delta.mjs, preserve the current selected lanes/scopes, but replace the duplicated downstream lane arrays with a smaller policy model: direct test owner per workspace, explicit non-manifest integration edges/overrides, and a computed reverse-dependency closure from trusted package manifests or generated trusted data. Prove with the existing snapshot/table tests that every current plan is unchanged.

File: .github/workflows/evm-integration.yml (hunk 3 above)
Comment (otReviewAgent) [reply .github/workflows/evm-integration.yml:50]: 🟡 Issue: Factor the duplicated planner bootstrap out of the workflows

**What's wrong**
The PR extracts planner logic into scripts, but leaves the orchestration that feeds it duplicated in YAML across the primary and EVM workflows. This keeps policy-critical control flow spread across two large workflow files, increasing the chance of future drift.

**Example**
If the base-vs-merge comparison, DELTA_ENABLED rollout rule, trusted sparse checkout, or planner arguments need to change, the maintainer has to patch both workflows in lockstep. A small skew would be hard to notice because both copies look like policy boilerplate rather than one owned implementation.

**Suggested direction**
Keep the trust boundary, but put the shared checkout/diff/PLAN_EVENT/planner invocation mechanics behind one reusable trusted entrypoint so the workflows only express their job graph.

**For Agents**
Move the duplicated planner bootstrap into one trusted script/composite action, for example a trusted-ci script that receives EVENT_NAME, BASE_SHA, MERGE_SHA, labels, sample key, and output paths, writes the name-status file, applies the delta-disabled event rewrite, and calls planCi. Keep workflow-specific outputs the same and verify both workflows still produce the current plan_json/evm_matrix behavior.

File: scripts/ci/plan-vitest-shard.mjs (hunk 8 above)
Comment (otReviewAgent) [reply scripts/ci/plan-vitest-shard.mjs:30]: 🟡 Issue: Separate sharding data from sharding logic

**What's wrong**
The shard planner mixes volatile benchmark data with the actual planning algorithm. The code that matters is only a small portion of the file, but it is buried between two large hard-coded timing maps tied to a specific Actions run.

**Example**
Renaming one CLI test file or refreshing timing data means editing a 450-line executable module mostly composed of historical measurement data. Reviewers must scan through hundreds of constants before reaching the actual discovery and bin-packing logic.

**Suggested direction**
Store the timing snapshots as data and load them from the planner, or generate them from a checked-in fixture. That would make algorithm changes and timing refreshes reviewable independently.

**For Agents**
Preserve the current shard assignments, but move CLI_BODY_MS and CHAIN_BODY_MS into data files such as scripts/ci/vitest-shard-weights/cli.json and chain.json, then keep plan-vitest-shard.mjs focused on discovery, validation, and bin-packing. The existing vitest-shards tests should assert the data files cover known files and keep output deterministic.

File: docs/ci-delta-policy.md [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+## Developer workflow
+
+- Open or update a PR normally. The **Plan CI delta** job summary lists selected
+  and skipped lanes and explains the reason.
+- Add `ci:full` when the change is riskier than its paths imply or when a
+  reviewer asks for the complete suite. Adding/removing the label reruns CI.
+- Use **Run workflow** for an unconditional full run on any branch.
+- If a new package, dependency edge, or integration consumer is introduced,
+  update `WORKSPACE_RULES` and its routing snapshot in the same PR. A package
+  manifest change already forces full CI for that PR.
```
Comment (lupuszr) [inline docs/ci-delta-policy.md:91]: ### [P2] Rotate the trusted controller with routing updates

Updating `WORKSPACE_RULES` alone does not activate the new policy because both planners and both aggregate gates execute from the immutable `c441bd6766ace67e331c0dfc6e22cb1325a6e1f6` controller commit. A manifest-changing PR does receive full CI, but after it merges, subsequent PRs still use the old dependency closure and can skip newly affected lanes during PR feedback.

Please document the required sequence: first commit the routing-policy change, then rotate all four trusted-controller `ref` values in `.github/workflows/ci.yml` and `.github/workflows/evm-integration.yml` to that reviewed commit. A regression check requiring policy changes and controller-pin rotation to stay paired would make this harder to miss.

The routing at the merged head itself is consistent and its completed CI runs are green; this finding concerns the documented procedure for future policy updates.
